gulpfile.babel.js

```diff
@@ -93,10 +98,80 @@ gulp.task('default', ['clean'], cb =>
     'lint',
     'vulcanize',
     'scripts',
+    'generate-service-worker',
     cb
   )
 );
 
+// Copy over the scripts that are used in importScripts as part of the generate-service-worker task.
+gulp.task('copy-sw-scripts', () => {
+  return gulp.src('node_modules/sw-toolbox/sw-toolbox.js')
+    .pipe(gulp.dest('static/dist'));
+});
+
+// Generate a service worker file that will provide offline functionality for
+// local resources.
+gulp.task('generate-service-worker', ['copy-sw-scripts'], () => {
+  const staticDir = 'static';
+  const distDir = path.join(staticDir, 'dist');
+  const filepath = path.join(distDir, 'service-worker.js');
+
+  // Precache all the truly static resources, and everything else will be
+  // cached at runtime. While the various Polymer elements are static, they are
+  // not pre-cached. Reasons include:
+  //  - The elements may be served in different ways. Typically, the Polymer
+  //    elements are vulcanized into a single file per page, but may be served
+  //    individually during development or testing.
+  //  - The vulcanized files end up being rather large. Rather than consuming
+  //    more space on service worker install, wait until the user visits the
+  //    including page
+  //  - The pages already include logic to preload/push the elements. Rely on
+  //    that logic to improve the initial loading performance of the page.
+  return swPrecache.write(filepath, {
+    cacheId: 'chromestatus',
+    verbose: true,
+    logger: $.util.log,
+    staticFileGlobs: [
+      // Images
+      `${staticDir}/img/**/*`,
+      `${staticDir}/elements/openinnew.svg`,
+      // Scripts
+      `${staticDir}/bower_components/webcomponentsjs/webcomponents-lite.min.js`,
+      `${staticDir}/js/**/*.js`,
+      // Styles
+      `${staticDir}/css/**/*.css`,
+    ],
+    runtimeCaching: [
+      // Server-side generated content
+      {
+        urlPattern: /\/features\/\w+$/,
+        handler: 'fastest',
+        options: {
+          cache: {
+            maxEntries: 10,
+            name: 'features-cache'
+          }
+        }
+      },
+      // For dynamic data (json), try the network first to get the most recent
+      // values.
+      {
+        urlPattern: /\/features.json$/,
+        handler: 'networkFirst',
+      },
+      {
+        urlPattern: /\/omaha_data$/,
+        handler: 'networkFirst',
+      },
+      // Polymer elements that may or may not be vulcanized
+      {
+        urlPattern: /\/static\/(bower_components|elements)\//,
+        handler: 'cacheFirst',
```

Comment:
Yes, it would be my preference to use sw-precache for all the included Polymer elements. However, I'm concerned about the size of the content - how much is reasonable? Adding the vulcanized files to sw-precache, adds over 1MB to be fetched on install (total precache size 117kB -> 1.5MB, as in the see the output below).

We could:
- Precache only feature-imports.vulcanized.*, as this is part of the main view, and will have to be downloaded/cached anyway to work offline
- Use runtime caching, strategy = fastest, for the other static elements.

Partial output from sw-precache for vulcanized content:
[11:57:36] Caching static resource "static/elements/admin-imports.vulcanize.html" (39.92 kB)
[11:57:36] Caching static resource "static/elements/admin-imports.vulcanize.js" (255.38 kB)
[11:57:36] Caching static resource "static/elements/features-imports.vulcanize.html" (74.05 kB)
[11:57:36] Caching static resource "static/elements/features-imports.vulcanize.js" (303.58 kB)
[11:57:36] Caching static resource "static/elements/metrics-imports.vulcanize.html" (124.71 kB)
[11:57:36] Caching static resource "static/elements/metrics-imports.vulcanize.js" (264.49 kB)
[11:57:36] Caching static resource "static/elements/samples-imports.vulcanize.html" (47.89 kB)
[11:57:36] Caching static resource "static/elements/samples-imports.vulcanize.js" (270.76 kB)
[11:57:36] Total precache size is about 1.5 MB for 36 resources.